### PR TITLE
Congestion next

### PR DESCRIPTION
**Pull request by Stolas**
we had a problem that the queue would lock if one queue is full. This patch switches when a queue is full if a specific config value is set (:congestion_switch) couple of lines so you can easily see my changes.

### CODE REVIEW

**Comment by Stolas** [comment]
Fixed, signed the document

File: lib/logstash/outputs/redis.rb [hunk 1]
```diff
@@ -92,6 +92,9 @@ class LogStash::Outputs::Redis < LogStash::Outputs::Base
   # Only supported for `list` Redis `data_type`.
   config :congestion_threshold, :validate => :number, :default => 0
 
+	# Auto-Queue switching when full, instead of waiting.
+  config :congestion_switch, :validate => :boolean, :default => false
```
Comment (jsvd) [inline lib/logstash/outputs/redis.rb]: This config option `congestion_switch` is not used in the code

File: lib/logstash/outputs/redis.rb [hunk 2]
```diff
@@ -185,7 +188,13 @@ def congestion_check(key)
     return if @congestion_threshold == 0
     if (Time.now.to_i - @congestion_check_times[key]) >= @congestion_interval # Check congestion only if enough time has passed since last check.
       while @redis.llen(key) > @congestion_threshold # Don't push event to Redis key which has reached @congestion_threshold.
-        @logger.warn? and @logger.warn("Redis key size has hit a congestion threshold #{@congestion_threshold} suspending output for #{@congestion_interval} seconds")
+        if @shuffle_hosts
+          # Switch to different server, and try again.
+          @logger.warn? and @logger.warn("Redis key size has hit a congrestion threshold #{@congestion_threshold} switching to next server.")
+          @redis = connect
```
Comment (jordansissel) [inline lib/logstash/outputs/redis.rb]: this should probably call @redis.quit to shutdown the previous connection.

File: lib/logstash/outputs/redis.rb [hunk 3]
```diff
@@ -185,7 +188,13 @@ def congestion_check(key)
...
+        else
+          @logger.warn? and @logger.warn("Redis key size has hit a congestion threshold #{@congestion_threshold} suspending output for #{@congestion_interval} seconds")
```
Comment (jordansissel) [inline lib/logstash/outputs/redis.rb]: Contextual logging (A message and the related data) should have the message and data be totally separate. For example:

```ruby
@logger.warn? && @logger.warn("Redis key size has hit congestion threshold. Suspending output temporarily", :threshold => @congestion_threshold, :interval => @congestion_interval)
```

File: lib/logstash/outputs/redis.rb [hunk 4]
```diff
@@ -185,7 +188,13 @@ def congestion_check(key)
     return if @congestion_threshold == 0
     if (Time.now.to_i - @congestion_check_times[key]) >= @congestion_interval # Check congestion only if enough time has passed since last check.
       while @redis.llen(key) > @congestion_threshold # Don't push event to Redis key which has reached @congestion_threshold.
-        @logger.warn? and @logger.warn("Redis key size has hit a congestion threshold #{@congestion_threshold} suspending output for #{@congestion_interval} seconds")
+        if @shuffle_hosts
+          # Switch to different server, and try again.
+          @logger.warn? and @logger.warn("Redis key size has hit a congrestion threshold #{@congestion_threshold} switching to next server.")
```
Comment (jordansissel) [inline lib/logstash/outputs/redis.rb]: See my contextual logging comment below; same comment here.

File: lib/logstash/outputs/redis.rb [hunk 5]
```diff
@@ -185,7 +188,13 @@ def congestion_check(key)
     return if @congestion_threshold == 0
     if (Time.now.to_i - @congestion_check_times[key]) >= @congestion_interval # Check congestion only if enough time has passed since last check.
       while @redis.llen(key) > @congestion_threshold # Don't push event to Redis key which has reached @congestion_threshold.
-        @logger.warn? and @logger.warn("Redis key size has hit a congestion threshold #{@congestion_threshold} suspending output for #{@congestion_interval} seconds")
+        if @shuffle_hosts
```
Comment (jordansissel) [inline lib/logstash/outputs/redis.rb]: Why would we do this only if shuffle_hosts is set? Did you mean `if @congestion_switch` ?

File: lib/logstash/outputs/redis.rb [hunk 6]
```diff
@@ -92,6 +92,9 @@ class LogStash::Outputs::Redis < LogStash::Outputs::Base
   # Only supported for `list` Redis `data_type`.
   config :congestion_threshold, :validate => :number, :default => 0
 
+	# Auto-Queue switching when full, instead of waiting.
```
Comment (jordansissel) [inline lib/logstash/outputs/redis.rb]: This doc is a bit confusing to me, here's an alternative:

"If multiple redis servers are configured and congestion is detected, switch to a different redis server."

**Comment by jordansissel** [comment]
While this is new behavior, I am thinking that we should make the "switch on congestion" enabled by default.

Any thoughts?

**Comment by Stolas** [comment]
I have made a really quick hack because a friend of me required this function. If I have made some errors to get upstream I would gladly help and fix this, but as it has been quite a while I can not even remember what I did anymore. 

Anyway, I will have a quick look at the code during the weekend and fix it up so it will have all the requirements and it can get upstream.

**Comment by jordansissel** [comment]
@sto

**Comment by jordansissel** [comment]
@Stolas No worries! We'll get this ironed out together :)

File: lib/logstash/outputs/redis.rb (hunk 6 above)
Comment (Stolas) [inline lib/logstash/outputs/redis.rb]: Sounds perfect

File: lib/logstash/outputs/redis.rb (hunk 1 above)
Comment (Stolas) [inline lib/logstash/outputs/redis.rb]: This should be the switch to determine if the switch should be done or not. (eg if this patch should be enabled or if old function should be used).

File: lib/logstash/outputs/redis.rb (hunk 5 above)
Comment (Stolas) [inline lib/logstash/outputs/redis.rb]: yes

File: lib/logstash/outputs/redis.rb (hunk 2 above)
Comment (Stolas) [inline lib/logstash/outputs/redis.rb]: Ok, I can add that.

**Comment by jordansissel** [comment]
For Logstash 1.5.0, we've moved all plugins (and grok patterns) to individual repositories. Can you move this pull request to https://github.com/logstash-plugins/logstash-output-redis?

This sequence of steps _may_ help you do this migration:

1. Fork this repository
2. Clone your fork

  ```shell
  git clone https://github.com/Stolas/logstash.git
  ```

3. Create a branch:

  ```shell
  git checkout -b my-branch-name
  ```

4. Apply the patches from this PR into your branch:

  ```shell
  curl -s https://github.com/elastic/logstash/pull/1616.patch | git am
  ```

  This should take your commits from this PR and commit them to your new local branch.

5. Push!

  ```shell
  git push origin my-branch-name
  ```

6. Open a new PR against https://github.com/logstash-plugins/logstash-output-redis

**Pull request closed without merging**